src/fs/cluster_allocator.cc

```diff
@@ -21,34 +21,86 @@
 
 #include "fs/cluster.hh"
 #include "fs/cluster_allocator.hh"
+#include "seastar/core/future.hh"
 
 #include <cassert>
 #include <optional>
 
-namespace seastar {
+namespace seastar::fs {
 
-namespace fs {
+cluster_allocator::cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters,
+        circular_buffer<cluster_id_t> free_clusters)
+        : _allocated_clusters(std::move(allocated_clusters))
+        , _free_clusters(std::move(free_clusters))
+        , _cluster_sem(_free_clusters.size()) {
+    size_t cluster_nb = _free_clusters.size() + _allocated_clusters.size();
+    _allocated_clusters.reserve(cluster_nb);
+    _free_clusters.reserve(cluster_nb);
+}
 
-cluster_allocator::cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters)
-        : _allocated_clusters(std::move(allocated_clusters)), _free_clusters(std::move(free_clusters)) {}
+void cluster_allocator::reset(std::unordered_set<cluster_id_t> allocated_clusters,
+        circular_buffer<cluster_id_t> free_clusters) {
+    assert(_cluster_sem.available_units() == _free_clusters.size());
+    assert(_cluster_sem.waiters() == 0);
 
-std::optional<cluster_id_t> cluster_allocator::alloc() {
-    if (_free_clusters.empty()) {
-        return std::nullopt;
-    }
+    size_t cluster_nb = free_clusters.size() + allocated_clusters.size();
+    allocated_clusters.reserve(cluster_nb);
+    free_clusters.reserve(cluster_nb);
+
+    _allocated_clusters = std::move(allocated_clusters);
+    _free_clusters = std::move(free_clusters);
+    _cluster_sem.consume(_cluster_sem.available_units());
+    _cluster_sem.signal(_free_clusters.size());
+}
+
+cluster_id_t cluster_allocator::do_alloc() noexcept {
+    assert(not _free_clusters.empty());
 
     cluster_id_t cluster_id = _free_clusters.front();
     _free_clusters.pop_front();
-    _allocated_clusters.insert(cluster_id);
+    _allocated_clusters.emplace(cluster_id);
```

Comment:
I am not sure that reserve on unordered_set makes emplace noexcept. Documentations says that it only reserves buckets, but I don't know what about allocating the list element on a hash collision (or even when there is no collision, but this is the first element in that bucket).

Reply:
https://en.cppreference.com/w/cpp/container/unordered_set
Here it says that reserve "reserves space for at least the specified number of elements. This regenerates the hash table." And for rehash "reserves at least the specified number of buckets. This regenerates the hash table." But that is the only place where I found it.

Reply:
You are right. I found that question.
https://stackoverflow.com/questions/42373232/why-does-stdunordered-map-have-a-reserve-method